MXB Pamphlet xmatter

gmartin7 opened the pull request:
* This includes a change to c# code so that
  Credit-Page-style will work on the copyright field
* Add customXMatter folder to templates
* Add jade and less to combine title/credit pages
* Add printer's statement space to inside back cover
* Rearrange some of the parts of the combined page
* Modify a few bits of less, so the rearranged bits fit
* Get Printer's Statement 'bottom aligned'
* Moved both acknowl fields up above funding
* Added Printing Info section just above copyright
* Reduced space below original acknowledgments
   and funding sections
* Reduced min-height to one line for funding and
   original contributions sections
* Get Credit-Page-style to work on all the combined
  title/credits page fields except the titles
* Maintain a separate style for the PrintingInfo
   section

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1272)

<!-- Reviewable:end -->

gmartin7 commented:
NB: This PR includes a change to C# code that allows the Copyright statement to benefit from the Credit-Page-style that it already has.

JohnThomson commented:
May be perfectly good, but some aspects really baffle me. Might be better to have it reviewed by someone with better knowledge of jade and the book-settings mechanism. It's quite tricky and easy to break...@Hatton has been gradually improving things, but I don't recall the overall strategy for doing that.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 364 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF6v_W_Ja2GlhDuGkw:-KWF6v_W_Ja2GlhDuGkx:b-q01n0t) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4d83710611ac09ebc62e9882c6629c0b64563345/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L364)):*
> ```Less
>     }
> 
>     #printingInfo .bloom-contentNational1 {
> ```

I like to see comments on display:inherit, because you tend to have to check a lot of rules to find out what actual behavior will be inherited and why there is more than one option and what might be messed up if you change this.  Are there two or more contexts where different things will be inherited? If not why is display:inherit better than setting the mode you want? (Or is this just following a pattern of a group of similar rules?)

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 460 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF7iL7pqxphTp0gYcZ:-KWF7iL7pqxphTp0gYc_:bdoelsd) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4d83710611ac09ebc62e9882c6629c0b64563345/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L460)):*
> ```Less
>         }
>     }
>     
> ```

spurious white space?

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 499 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF86S7UqZXERUexBsE:-KWF86S7UqZXERUexBsF:bs6uz6d) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4d83710611ac09ebc62e9882c6629c0b64563345/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L499)):*
> ```Less
>         display: flex;
>         flex-direction: column;
>     }
> ```

The trick you're doing here will need to be done, probably another way, for e-books, unless we decide to rule out making e-books with this x-matter. e-books use their own much-simplified style-sheet, and IIRC I found that relying on flex-order to arrange blocks vertically did not work (at least in some readers), so the export code has some smarts already to actually put the blocks in the right order.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 504 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF8zNhZ-QqbkjwJp67:-KWF8zNiSWslRRVEy9_J:b-jdox7t) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4d83710611ac09ebc62e9882c6629c0b64563345/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L504)):*
> ```Less
> .outsideBackCover .bloom-editable {
>     display: inherit;
>     height: inherit; // let this shrink to make room for branding at bottom, if present
> ```

What value for height is it supposed to inherit that will achieve this goal? (Sorry, I know you didn't change this line...)

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade, line 161 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF9zGUzD7wqHsMhL4g:-KWF9zGVv-dU90pFtdAa:b-n6dh8w) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4d83710611ac09ebc62e9882c6629c0b64563345/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade#L161)):*
> ```Jade
> 
> mixin standard-combinedTitleCreditsPage-inside-front-cover
> 	+page-xmatter('Inside Front Cover').cover.coverColor.titlePage.credits.bloom-frontMatter(data-export='front-matter-title-page')&attributes(attributes)#0d61e568-6814-4836-82ad-81b2bcd106a5
> ```

It surprises me to see what I presume is the GUID for the page type history in an area I thought would be content...isn't it supposed to become an attribute of the page div? (May just be my lack of knowledge of jade)

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 114 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWFAdF9M_NOvttPtGW3:-KWFAdFAMI-xooeaIpGX:bej9tgn) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4d83710611ac09ebc62e9882c6629c0b64563345/src/BloomExe/Book/BookCopyrightAndLicense.cs#L114)):*
> ```Smalltalk
> 			dom.SetBookSetting("copyright", "*", metadata.CopyrightNotice);
> 			// This allows Credit-Page-style to "take" on the displayed copyright
> 			dom.SetBookSetting("copyright", collectionSettings.Language2Iso639Code, metadata.CopyrightNotice);
> ```

I have a suspicion that bad things will result from trying to store the copyright as both * and a particular language. I recall vaguely a bug where the (no-longer-used) 'en' version of something would keep coming back when the user tried to delete the * version entirely. We either know the language(s) the copyright notice is in, in which case we should store one or more versions so labeled, or we don't, in which case we should store a single * version. If there's a good reason to do both I think it needs more explanation (I'm not clear what the comment above means, let alone how having the duplicate copyright info might achieve it). Also carefully test deleting this string after it is saved both ways (also in a derived book which starts out with it in * and also a language not used in the book).
I suppose you've looked into how this inter-operates with storing copyright in clearshare metadata, since I think that's what metadata.CopyrightNotice is? I didn't think that had a way to store multiple-language copyright...this might get into complications with what can be stored in standard image metadata? But looks like you're not trying to store two versions there...in fact the only way you could have _different_ copyright strings would be if copyright is changed after language 2 is changed, which _may_ not be possible...but what if you 'unlock' a derived book?
Sorry, not sure all these complications will actually be problems, but seemed worth raising questions.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 162 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWFDWDZl7LbgOgk80lR:-KWFDWDZl7LbgOgk80lS:b-cbj4hz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/4d83710611ac09ebc62e9882c6629c0b64563345/src/BloomExe/Book/BookCopyrightAndLicense.cs#L162)):*
> ```Smalltalk
> 		internal static void UpdateDomFromDataDiv(HtmlDom dom, string bookFolderPath, CollectionSettings collectionSettings)
> 		{
> 			// Adding L2(N1) as first preference allows Credit-Page-style to "take" on the displayed copyright
> ```

Again, I don't know what it means 'to "take" on the displayed copyright'. Presumably we want the page to show the copyright the user entered into the dialog. What about that requires us to store the same thing two ways? And if it is the same thing stored both ways, why do we need a trick style to prefer one over the other?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-:-KWFETqG4psPpqToQRCs:b-5svahi)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Thanks for your input from far away! Some of this is not clear to me either, so I agree with you that @hatton should have a look... someday. I presume he's waiting until the changes settle down too. I'm still making small tweaks for the user, but I think I'm nearing completion.

---

Review status: 1 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 364 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF6v_W_Ja2GlhDuGkw:-KWdIPaLVTYKkhxM8xcY:bp2i4pv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L364)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> I like to see comments on display:inherit, because you tend to have to check a lot of rules to find out what actual behavior will be inherited and why there is more than one option and what might be messed up if you change this.  Are there two or more contexts where different things will be inherited? If not why is display:inherit better than setting the mode you want? (Or is this just following a pattern of a group of similar rules?)
</details>
'display: block;' worked just as well. I changed it. We just need something here to override basePage's 'display: none;'.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 460 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF7iL7pqxphTp0gYcZ:-KWd5l-fqsAo9VSXeuhx:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L460)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> spurious white space?
</details>
Done.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 499 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF86S7UqZXERUexBsE:-KWd5txFg6F4_s0b2LR8:b-5gsavn) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L499)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> The trick you're doing here will need to be done, probably another way, for e-books, unless we decide to rule out making e-books with this x-matter. e-books use their own much-simplified style-sheet, and IIRC I found that relying on flex-order to arrange blocks vertically did not work (at least in some readers), so the export code has some smarts already to actually put the blocks in the right order.
</details>
Trying to see if removing the 'flex-order' elements will affect anything... doesn't seem to. 'flex-order's removed.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 504 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF8zNhZ-QqbkjwJp67:-KWdCVyhZGnootflBvQN:b-wkcm0n) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L504)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> What value for height is it supposed to inherit that will achieve this goal? (Sorry, I know you didn't change this line...)
</details>
Sorry. I don't know the answer to that. But I do see that I (accidentally) reverted a change that @hatton made recently to made 'display: block -> inherit'. I'll change that back (unless I see it causes problems). I think I was trying to get it to match up with the .insideBackCover value, but it has 'height: 100%;' so it may need a different value for 'display'.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade, line 161 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF9zGUzD7wqHsMhL4g:-KWd3llKflnWKSDQH2UM:bhan4pm) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade#L161)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> It surprises me to see what I presume is the GUID for the page type history in an area I thought would be content...isn't it supposed to become an attribute of the page div? (May just be my lack of knowledge of jade)
</details>
Yeah, I think jade/pug has changed significantly since I was working in it originally, but this guid just shows up as the page id. Each new type of page in this file has its own guid set the same way.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 114 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWFAdF9M_NOvttPtGW3:-KWdAWSwI6dx-819_0d1:b45uguk) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomExe/Book/BookCopyrightAndLicense.cs#L114)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> I have a suspicion that bad things will result from trying to store the copyright as both * and a particular language. I recall vaguely a bug where the (no-longer-used) 'en' version of something would keep coming back when the user tried to delete the * version entirely. We either know the language(s) the copyright notice is in, in which case we should store one or more versions so labeled, or we don't, in which case we should store a single * version. If there's a good reason to do both I think it needs more explanation (I'm not clear what the comment above means, let alone how having the duplicate copyright info might achieve it). Also carefully test deleting this string after it is saved both ways (also in a derived book which starts out with it in * and also a language not used in the book).
> I suppose you've looked into how this inter-operates with storing copyright in clearshare metadata, since I think that's what metadata.CopyrightNotice is? I didn't think that had a way to store multiple-language copyright...this might get into complications with what can be stored in standard image metadata? But looks like you're not trying to store two versions there...in fact the only way you could have _different_ copyright strings would be if copyright is changed after language 2 is changed, which _may_ not be possible...but what if you 'unlock' a derived book?
> Sorry, not sure all these complications will actually be problems, but seemed worth raising questions.
</details>
See below for my attempt to answer this. I've also modified the comments to make it more clear what I mean by a style 'take'-ing. Also, unless you're talking about the user messing directly with the .html file, you can no longer remove the Copyright once it's entered! You can modify it, but if you delete it the OK button on the dialog in disabled.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 162 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWFDWDZl7LbgOgk80lR:-KWd85W_EULyjcf7oR6v:b-isq7sv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomExe/Book/BookCopyrightAndLicense.cs#L162)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> Again, I don't know what it means 'to "take" on the displayed copyright'. Presumably we want the page to show the copyright the user entered into the dialog. What about that requires us to store the same thing two ways? And if it is the same thing stored both ways, why do we need a trick style to prefer one over the other?
</details>
Credit-Page-style is a class that has been applied to the LicenseCopyright block for some time (indeed to everything on the credits page), but no styles associated with that class ever (up 'til now) applied themselves to the Copyright statement. (That's what I mean by not 'taking'.) With this change, suddenly the Copyright statement style obeys what is in the Credit-Page-style (either the default style value or a user overridden value). I tested it some to make sure both: 1) the user-entered copyright didn't get lost, and 2) any national lang. Credit-Page-style values applied to the copyright. If we don't store the L2 version, the styles don't apply, if we don't store the '*' version, the data can get lost. So far I haven't seen the Copyright line duplicated or anything bizarre like that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-:-KWd4OjNoTNNiYr_klvT:bqclq2e)*
<!-- Sent from Reviewable.io -->

hatton commented:
Reviewed 1 of 6 files at r1, 2 of 5 files at r2.
Review status: 3 of 6 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.jade, line 16 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7ZsDsYrVdWQa20ERz:-KX7ZsDtUGNfCxbQfMKA:ba5fh1d) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.jade#L16)):*
> ```Jade
> 		+factoryStandard-outsideFrontCover
> 		+standard-combinedTitleCreditsPage-inside-front-cover
> 		+standard-nationalSummaryprinterStatement-insideBackCover
> ```

I'm not clear that Printer's Statement rises to the level of "standard" yet. When it becomes standard to include a printer's statement, it will be in all the xmatter choices. So that should be in this file.

---

*[src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.less, line 27 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7bh_nk6sEMRRNfJQT:-KX7bh_o5ZuboU5BMGks:bkn7qfw) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.less#L27)):*
> ```Less
> 
> // default values for the printInfo section
> .PrintingInfo-style.bloom-contentNational1 {
> ```

PrintingInfo --> PrintersStatement

---

*[src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.less, line 41 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7bZg_JepfShArfd_J:-KX7bZg_JepfShArfd_K:b-z5fzae) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.less#L41)):*
> ```Less
> }
> 
> BODY[bookcreationtype="original"] {
> ```

I can't tell what the difference is between this and the next (translation) block.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 360 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7_M0zyoyjoK1RZ4iz:-KX7_M1-I6yX86nt9sOo:b7ukrh1) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L360)):*
> ```Less
>         }
>     }
>     #printingInfo .bloom-contentNational1 {
> ```

Please move to an MX-specific stylesheet. I think "printersStatement" would be a better name.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 468 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7cAyoG7U6zGkIPj0l:-KX7cAyoG7U6zGkIPj0m:b-fomnz0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L468)):*
> ```Less
> 
> .insideBackCover {
>     .marginBox {
> ```

Please move these changes to mx-specific less file.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade, line 160 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7b4gqGb4O5nK5HnU0:-KX7b4gqGb4O5nK5HnU1:bacfj9m) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade#L160)):*
> ```Jade
> 		+title-page-contents
> 
> mixin standard-combinedTitleCreditsPage-inside-front-cover
> ```

Please move to mx-specific jade for now

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade, line 180 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7ayhKj6U1ZYaoy7K0:-KX7ayhKj6U1ZYaoy7K1:b-i6sh0g) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade#L180)):*
> ```Jade
> 		+field-prototypeDeclaredExplicity#printingInfo
> 			label.bubble Use this for Printing Info (1st Edition, etc.)
> 			+editable(kLanguageForPrototypeOnly).printInfo.PrintingInfo-style.bloom-copyFromOtherLanguageIfNecessary(data-book='printingInfo')
> ```

printingInfo-->printersStatement

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade, line 219 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7ahbVpxNcPNQJ9_rB:-KX7ahbVpxNcPNQJ9_rC:bwz69xj) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade#L219)):*
> ```Jade
> 			label.bubble If you need somewhere to put more information about the book, you can use this page, which is the inside of the back cover.
> 
> mixin standard-nationalSummaryprinterStatement-insideBackCover
> ```

Please remove "standard" and move to mx-specific jade. And "p" wants to be capitalized.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 114 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWFAdF9M_NOvttPtGW3:-KX7aYGXK61l3mqz6PDN:b4jmtda) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomExe/Book/BookCopyrightAndLicense.cs#L114)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary>
> See below for my attempt to answer this. I've also modified the comments to make it more clear what I mean by a style 'take'-ing. Also, unless you're talking about the user messing directly with the .html file, you can no longer remove the Copyright once it's entered! You can modify it, but if you delete it the OK button on the dialog in disabled.
</details>
Yah, the implications of this are too murky for me, too. For now, can you split changes to this file off as a separate PR? Thx.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Review status: 3 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 499 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF86S7UqZXERUexBsE:-KX8Fm08lt6Qa9kAy9Xw:bg2o5el) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L499)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary>
> Trying to see if removing the 'flex-order' elements will affect anything... doesn't seem to. 'flex-order's removed.
</details>
That's good...I'd still recommend that making an epub should be part of the testing for a new xmatter pack.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 162 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWFDWDZl7LbgOgk80lR:-KX8GnOGTOSsh3_wrBxi:b-8i8smz) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomExe/Book/BookCopyrightAndLicense.cs#L162)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary>
> Credit-Page-style is a class that has been applied to the LicenseCopyright block for some time (indeed to everything on the credits page), but no styles associated with that class ever (up 'til now) applied themselves to the Copyright statement. (That's what I mean by not 'taking'.) With this change, suddenly the Copyright statement style obeys what is in the Credit-Page-style (either the default style value or a user overridden value). I tested it some to make sure both: 1) the user-entered copyright didn't get lost, and 2) any national lang. Credit-Page-style values applied to the copyright. If we don't store the L2 version, the styles don't apply, if we don't store the '*' version, the data can get lost. So far I haven't seen the Copyright line duplicated or anything bizarre like that.
</details>
So you're saying somewhere there's a rule that fires on elements with the Credit-Page-Style and lang=<lang2> but not on ones with that style and lang="*"? And you're adding a more-or-less redundant element to get the rule to fire? Seems something else must be different, because AFAIK our style sheets don't know about actual languages, only styles like bloom-contentNational1 and so on.
I think this deserves more research. Surely there's a way to make the rule you want match the lang="*" copyright element? Even if you had to add another match condition to the rule, I think that would be safer than adding a duplicate copyright notice which could so easily get out of sync with different versions of the programs manipulating files? If there really isn't a better way I think comments should spell out the problem and why this is the best solution.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: 0 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.jade, line 16 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7ZsDsYrVdWQa20ERz:-KXC8l_yjlM2ZxJIhHHP:bhmyckv) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.jade#L16)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> I'm not clear that Printer's Statement rises to the level of "standard" yet. When it becomes standard to include a printer's statement, it will be in all the xmatter choices. So that should be in this file.
</details>
renamed both to mxb-*

---

*[src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.less, line 27 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7bh_nk6sEMRRNfJQT:-KXBykho3dbmTcPd_eNx:b-pau34l) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.less#L27)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> PrintingInfo --> PrintersStatement
</details>
changed to PrintingHistory

---

*[src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.less, line 41 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7bZg_JepfShArfd_J:-KX7gfLzswjp2HC4OUBx:bc7dtu) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/customXMatter/MXBPamphlet-XMatter/MXBPamphlet-XMatter.less#L41)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> I can't tell what the difference is between this and the next (translation) block.
</details>
The 'original' uses .bloom-content1, the 'translation' uses .bloom-contentNational1. I added comments to make it clearer.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 499 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWF86S7UqZXERUexBsE:-KXC9v-jWUiUuf9yBB9Q:b-i9e98r) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L499)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> That's good...I'd still recommend that making an epub should be part of the testing for a new xmatter pack.
</details>
Hmm... yes. Making an epub results in printer's statement not bottom justified. Nothing wrong with the order, though.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 360 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7_M0zyoyjoK1RZ4iz:-KXByr6e9OP4-zYAXmIj:b5qleb4) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L360)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Please move to an MX-specific stylesheet. I think "printersStatement" would be a better name.
</details>
changed to #printingHistory

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less, line 468 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7cAyoG7U6zGkIPj0l:-KXC9hHo7rhEuN4PmMJj:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-common.less#L468)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Please move these changes to mx-specific less file.
</details>
Done.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade, line 160 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7b4gqGb4O5nK5HnU0:-KXC9k8EIxRcSnHZeTOk:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade#L160)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Please move to mx-specific jade for now
</details>
Done.

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade, line 180 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7ayhKj6U1ZYaoy7K0:-KXBz48XAfrHRBASZhmq:b-8t8jed) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade#L180)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> printingInfo-->printersStatement
</details>
changed to printingHistory

---

*[src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade, line 219 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KX7ahbVpxNcPNQJ9_rB:-KXC9istnwrYn8mSNiAY:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/aeb59f67aaceaf83ae46c57567c0a77108c01b0b/src/BloomBrowserUI/templates/xMatter/bloom-xmatter-mixins.jade#L219)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Please remove "standard" and move to mx-specific jade. And "p" wants to be capitalized.
</details>
Done.

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 114 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWFAdF9M_NOvttPtGW3:-KXBzAFgN7pqJUSP1yDR:b9xegj3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomExe/Book/BookCopyrightAndLicense.cs#L114)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary>
> Yah, the implications of this are too murky for me, too. For now, can you split changes to this file off as a separate PR? Thx.
</details>
Done (and comment thread moved too).

---

*[src/BloomExe/Book/BookCopyrightAndLicense.cs, line 162 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272#-KWFDWDZl7LbgOgk80lR:-KXBzD4KCAuuZ6iKnDAT:by0hmgp) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/c2048c6641da787873905c98a2db350397a08f29/src/BloomExe/Book/BookCopyrightAndLicense.cs#L162)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> So you're saying somewhere there's a rule that fires on elements with the Credit-Page-Style and lang=<lang2> but not on ones with that style and lang="*"? And you're adding a more-or-less redundant element to get the rule to fire? Seems something else must be different, because AFAIK our style sheets don't know about actual languages, only styles like bloom-contentNational1 and so on.
> I think this deserves more research. Surely there's a way to make the rule you want match the lang="*" copyright element? Even if you had to add another match condition to the rule, I think that would be safer than adding a duplicate copyright notice which could so easily get out of sync with different versions of the programs manipulating files? If there really isn't a better way I think comments should spell out the problem and why this is the best solution.
</details>
JT's comment attached to separate PR with this C# code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1272)*
<!-- Sent from Reviewable.io -->
